File: chapter1/strategymode.py
#!/usr/env python


class Flyable:
    def fly(self):
        pass

class Quackable(object):
    def quack(self):
        pass

class ReadHeadDuckFly(Flyable):
    def fly(self):
        print "I am a readheadduck, I can fly"

class ReadHeadDuckQack(Quackable):
    def quack(self):
        print "I am a readheadduck,Dcuk duck duck..."

class Duck():
    def swim(self):
        print "I am a duck,I can swim..."

class ReadHeadDuck(Duck):
    def __init__(self,flyable,quackable):
        self.f = flyable
        self.q = quackable

    def fly(self):
        return self.f.fly()
    def quack(self):
        return self.q.quack()

class Mallardduckflyable(Flyable):
    def fly(self):
        print "I am a Mallardduck....,I can fly"

class MallardduckQuackble(Quackable):
    def quack(self):
        print "I am a Mallardduck,Duck.duck..duck.."

class Mallardduck(Duck):
    def __init__(self,flyable,quackable):
        self.f = flyable
        self.q = quackable

    def fly(self):
        return self.f.fly()

    def quack(self):
        return self.q.quack()

if __name__ == "__main__":
    duck = Duck()
    duck.swim()
    
    rhduck = ReadHeadDuck(ReadHeadDuckFly(),ReadHeadDuckQack())
    rhduck.fly()
    rhduck.swim()
    rhduck.quack()

    md = Mallardduck(Mallardduckflyable(),MallardduckQuackble())
    md.fly()
    md.quack()
    md.swim()

File: chapter1/strategymode.cc
#include<iostream>
using namespace std;

class FlyBehavior {
    public :
        virtual  void fly(void) = 0;
};

class Duck {
    public :
        FlyBehavior *flyBehavior;
        Duck() {}
        ~Duck() {}
        void performfly(void) {
            flyBehavior->fly();
        }
        void Swim(void) {
            cout <<"I am a duck, I can swim \n";
        }
};

class FlyWithWings : public FlyBehavior {
    public :
        void fly(void) {
            cout << "I'm fly with wings\n";
        }

};

class FlyNoWay : public FlyBehavior {
    public :
        void fly(void) {
            cout << "I can't fly \n";
        }
};

class RedHeadDuck : public Duck {
    public :
        RedHeadDuck(FlyBehavior *flyway) {
            flyBehavior = flyway;
        };
        void performfly(void) {
            flyBehavior->fly();
        }
};

class MallardDuck : public Duck {
    public:
        MallardDuck(FlyBehavior *flynoway) {
            flyBehavior = flynoway;
        }
        void performfly(void) {
           flyBehavior->fly();
        }
};
int main(void) {
    //Duck *duck = new Duck(new FlyWithWings() );
    Duck *duck = new Duck();
    duck->Swim();

    Duck *rd = new RedHeadDuck(new FlyWithWings() );
    rd->performfly();

    Duck *md = new MallardDuck(new FlyNoWay() );
    md->performfly();
    return 1;
}
